### PR TITLE
MAILBOX-381 DeletedMessagesVault implementation on top of a MailRepository

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    private final MailboxId.Factory mailboxIdFactory;
+    private final MessageId.Factory messageIdFactory;
+
+    public MailAdapter(MailboxId.Factory mailboxIdFactory, MessageId.Factory messageIdFactory) {
+        this.mailboxIdFactory = mailboxIdFactory;
+        this.messageIdFactory = messageIdFactory;
+    }
+
+    Mail toMail(DeletedMessage deletedMessage) throws MessagingException {
+        MimeMessageCopyOnWriteProxy mimeMessage = new MimeMessageCopyOnWriteProxy(new MimeMessageInputStreamSource(deletedMessage.getMessageId().serialize(), deletedMessage.getContent().get()));
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: oh, please, don't propagate this craziness in new code.

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+    private final MailboxId.Factory mailboxIdFactory;
+    private final MessageId.Factory messageIdFactory;
+
+    public MailAdapter(MailboxId.Factory mailboxIdFactory, MessageId.Factory messageIdFactory) {
+        this.mailboxIdFactory = mailboxIdFactory;
+        this.messageIdFactory = messageIdFactory;
+    }
+
+    Mail toMail(DeletedMessage deletedMessage) throws MessagingException {
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: this method is too long, please split every step into another method

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+            .collect(Guavate.toImmutableList());
+
+        AttributeValue<?> subjectAttributeValue = AttributeValue.of(deletedMessage.getSubject().map(AttributeValue::of));
+
+        return MailImpl.builder()
+            .name(deletedMessage.getMessageId().serialize())
+            .sender(deletedMessage.getSender())
+            .recipients(deletedMessage.getRecipients())
+            .mimeMessage(mimeMessage)
+            .attribute(new Attribute(OWNER_ATTRIBUTE_NAME, AttributeValue.of(new SerializableUser(deletedMessage.getOwner()))))
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: couldn't we simplify this Attribute syntax? it looks so ugly

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+import org.apache.mailet.AttributeName;
+import org.apache.mailet.AttributeValue;
+import org.apache.mailet.Mail;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+public class MailAdapter {
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: you only have package private methods, why do you need a public class and a public constructor ?

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+            .attribute(new Attribute(HAS_ATTACHMENT_ATTRIBUTE_NAME, AttributeValue.of(deletedMessage.hasAttachment())))
+            .attribute(new Attribute(DELIVERY_DATE_ATTRIBUTE_NAME, AttributeValue.of(new SerializableDate(deletedMessage.getDeliveryDate()))))
+            .attribute(new Attribute(DELETION_DATE_ATTRIBUTE_NAME, AttributeValue.of(new SerializableDate(deletedMessage.getDeletionDate()))))
+            .attribute(new Attribute(ORIGIN_MAILBOXES_ATTRIBUTE_NAME, AttributeValue.of(serializedMailboxIds)))
+            .attribute(new Attribute(SUBJECT_ATTRIBUTE_VALUE, subjectAttributeValue))
+            .build();
+    }
+
+    DeletedMessage fromMail(Mail mail) {
+        MemoizedSupplier<byte[]> byteContent = new MemoizedSupplier<>(Throwing.supplier(
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: do you expect to have callers not reading the message content?

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+            () -> MimeMessageUtil.asBytes(mail.getMessage())));
+
+        return DeletedMessage.builder()
+            .messageId(messageIdFactory.fromString(mail.getName()))
+            .originMailboxes(retrieveMailboxIds(mail))
+            .user(retrieveOwner(mail))
+            .deliveryDate(retrieveDate(mail, DELIVERY_DATE_ATTRIBUTE_NAME))
+            .deletionDate(retrieveDate(mail, DELETION_DATE_ATTRIBUTE_NAME))
+            .sender(mail.getMaybeSender())
+            .recipients(ImmutableList.copyOf(mail.getRecipients()))
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: the copy is the responsibility of the builder.

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+            .build();
+
+    }
+
+    private Optional<String> retrieveSubject(Mail mail) {
+        return mail.getAttribute(SUBJECT_ATTRIBUTE_VALUE)
+                .map(Attribute::getValue)
+                .map(AttributeValue::getValue)
+                .filter(value -> value instanceof Optional)
+                .map(value -> (Optional) value)
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: please use Optional.class::cast when doing cast

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+                .map(optional -> optional.map(obj -> {
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: a filter + cast would allow to reuse the `orElseThrow` below

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+            .recipients(ImmutableList.copyOf(mail.getRecipients()))
+            .content(() -> new ByteArrayInputStream(byteContent.get()))
+            .hasAttachment(retrieveHasAttachment(mail))
+            .subject(retrieveSubject(mail))
+            .build();
+
+    }
+
+    private Optional<String> retrieveSubject(Mail mail) {
+        return mail.getAttribute(SUBJECT_ATTRIBUTE_VALUE)
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: Attribute should provide a `<T> Optional<T> getAttributeValueAs(AttributeName name, Class<T> type)` method to avoid this kind of ugly code (that will happen over and over again). WDYT?

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+                    Preconditions.checkArgument(obj instanceof String, "mail should have a 'subject' attribute being of type 'Optional<String>");
+                    return (String) obj;
+                }))
+                .orElseThrow(() -> new IllegalArgumentException("mail should have a 'subject' attribute being of type 'Optional<String>"));
+    }
+
+    private boolean retrieveHasAttachment(Mail mail) {
+        Attribute attribute = mail.getAttribute(HAS_ATTACHMENT_ATTRIBUTE_NAME).orElseThrow(() -> new IllegalArgumentException("mail should have a 'hasAttachment' attribute"));
+        Preconditions.checkArgument(attribute.getValue().value() instanceof Boolean, "mail should have a 'hasAttachment' attribute of type boolean");
+        return (boolean) (Boolean) attribute.getValue().value();
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: that's definitely too ugly for me

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/SerializableDate.java [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+package org.apache.james.vault;
+
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Optional;
+
+import org.apache.mailet.ArbitrarySerializable;
+import org.apache.mailet.AttributeValue;
+
+public class SerializableDate implements ArbitrarySerializable<SerializableDate> {
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/SerializableDate.java]: you should provide a factory method that create returns a `AttributeValue<SerializableDate>` directly.

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/SerializableDate.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        public Optional<SerializableDate> deserialize(Serializable<SerializableDate> serializable) {
+            return Optional.of(serializable.getValue().value())
+                .filter(String.class::isInstance)
+                .map(String.class::cast)
+                .map(serializedDate -> ZonedDateTime.parse(serializedDate, DATE_TIME_FOMATTER))
+                .map(SerializableDate::new);
+        }
+    }
+
+    private static DateTimeFormatter DATE_TIME_FOMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ssZ");
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/SerializableDate.java]: what is specific about this pattern? Couldn't we use standard formatter instead?

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+import org.reactivestreams.Publisher;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.base.Preconditions;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+public class MailRepositoryBasedDeletedMessageVault implements DeletedMessageVault {
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java]: we never used `Based` for that purpose in class name. Could you elaborate on this choice?

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    private final MailAdapter mailAdapter;
+
+    public MailRepositoryBasedDeletedMessageVault(MailRepositoryStore mailRepositoryStore, Configuration configuration, MailAdapter mailAdapter) {
+        this.mailRepositoryStore = mailRepositoryStore;
+        this.configuration = configuration;
+        this.mailAdapter = mailAdapter;
+    }
+
+    Mono<MailRepository> repositoryForUser(User user) {
+        MailRepositoryUrl mailRepositoryUrl = MailRepositoryUrl.from(configuration.urlPrefix + user.asString());
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java]: I don't really like this string concatenation, couldn't we have a method for composing prefix and user?

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+        return Mono.fromCallable(() -> mailRepositoryStore.select(mailRepositoryUrl));
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java]: why do we need to return a Mono?

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    @Override
+    public Publisher<Void> append(User user, DeletedMessage deletedMessage) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(deletedMessage);
+
+        return repositoryForUser(user)
+            .flatMap(mailRepository -> Mono.fromCallable(() -> mailRepository.store(mailAdapter.toMail(deletedMessage))))
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java]: why do you put store operation into a callable? As you subscribe on the current thread, It will be executed synchronously anyway AFAIU

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+    @Override
+    public Publisher<Void> delete(User user, MessageId messageId) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(messageId);
+
+        MailKey mailKey = new MailKey(messageId.serialize());
+
+        return repositoryForUser(user)
+            .publishOn(Schedulers.elastic())
+            .flatMap(repository -> Mono.fromRunnable(
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java]: can't you run that with a simple map?

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+    @Override
+    public Publisher<DeletedMessage> search(User user, Query query) {
+        Preconditions.checkNotNull(user);
+        Preconditions.checkNotNull(query);
+        Preconditions.checkArgument(query.getCriteria().isEmpty(), "Search is not supported yet...");
+
+        Mono<MailRepository> mailRepositoryMono = repositoryForUser(user).cache();
+
+        return mailRepositoryMono
+            .flatMapMany(Throwing.function(repository -> Flux.fromStream(Iterators.toStream(repository.list()))))
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java]: can't you use `Flux.fromIterable` ?

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+            .flatMap(mailKey -> mailRepositoryMono.flatMap(repository -> Mono.fromCallable(() -> repository.retrieve(mailKey))))
+            .map(mailAdapter::fromMail)
+            .publishOn(Schedulers.elastic());
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java]: Why do you publishOn as the last step?

File: server/mailrepository/deleted-messages-vault-repository/src/test/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVaultTest.java [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+        Configuration configuration = Configuration.builder()
+            .workingDirectory("../")
+            .configurationFromClasspath()
+            .build();
+        FileSystemImpl fileSystem = new FileSystemImpl(configuration.directories());
+
+        MemoryMailRepositoryUrlStore urlStore = new MemoryMailRepositoryUrlStore();
+        MemoryMailRepositoryStore mailRepositoryStore = new MemoryMailRepositoryStore(urlStore, Sets.newHashSet(new MemoryMailRepositoryProvider()));
+        mailRepositoryStore.configure(new FileConfigurationProvider(fileSystem, configuration)
+            .getConfiguration("mailrepositorystore"));
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/test/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVaultTest.java]: couldn't we avoid that?

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessage.java [hunk 21]
```diff
@@ -145,6 +145,11 @@ public FinalStage subject(String subject) {
                 return this;
             }
 
+            public FinalStage subject(Optional<String> subject) {
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessage.java]: This is optional because a subject field might not be present in the headers.

If so, the best way to represent it is an optional...

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessage.java [hunk 22]
```diff
@@ -251,8 +251,7 @@ public final boolean equals(Object o) {
 
     @Override
     public final int hashCode() {
-        return Objects.hash(messageId, originMailboxes, owner, content, deliveryDate, deletionDate, sender, recipients,
-            subject, hasAttachment);
+        return Objects.hash(messageId, originMailboxes, owner, deliveryDate, deletionDate, sender, recipients, subject, hasAttachment);
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessage.java]: Well, writing the contract failed because equality between defferent kind of input streams is not handled by `Objects::equals`

I wrote a byte array equality stuff there but it ended up being not convincing either:
 - error management in equals and hashCode felt wrong
 - EqualsVerifier complained.

Then I went for the "less bad" option: testing content separately, like we do for every other classes.

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java (hunk 3 above)
Comment (chibenwa) [reply server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: I'd be glad to do it...

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java (hunk 5 above)
Comment (chibenwa) [reply server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: Yes I do. IE searches on date ranges. etc...

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java (hunk 9 above)
Comment (chibenwa) [reply server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: I would 100% upvote such a proposition (that I already made 6 months ago BTW)

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java (hunk 10 above)
Comment (chibenwa) [reply server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: Sorry, I did not write it consciously....

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java (hunk 13 above)
Comment (chibenwa) [reply server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java]: I did not take long to think about this name, to be honest.

Maybe just `MailRepositoryDeletedMessageVault` would be enough?

**Comment by Arsnael** [comment]
Read it...

**Comment by chibenwa** [comment]
We need tests covering corner cases for `Mail -> DeletedMessages` conversion...

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java (hunk 9 above)
Comment (chibenwa) [reply server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: I did contribute this, but as a method of AttributeValue

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java (hunk 8 above)
Comment (chibenwa) [reply server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: Not really because an empty option (no subject) is valid

**Comment by trantienduchn** [comment]
> Read it...

me too

**Comment by chibenwa** [comment]
We clearly have two concerns here that are quite independant:
 - Mail <-> DeletedMessage conversion: https://github.com/linagora/james-project/pull/2194
 - MailRepositoryDeletedMessageVault will also have a dedicated PR

File: mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessage.java (hunk 22 above)
[the remark this answers is not in the document]
Comment (mbaechler) [reply mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/vault/DeletedMessage.java]: You can very well compute a hash of the stream at construction time (and enforce it's immutable). That way, you don't require to handle errors in hashCode/equals.
Otherwise, I suggest you don't provide equality has an object property but instead rely on an external Comparator/Equality implementation.

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+
+        return DeletedMessage.builder()
+            .messageId(messageIdFactory.fromString(mail.getName()))
+            .originMailboxes(retrieveMailboxIds(mail))
+            .user(retrieveOwner(mail))
+            .deliveryDate(retrieveDate(mail, DELIVERY_DATE_ATTRIBUTE_NAME))
+            .deletionDate(retrieveDate(mail, DELETION_DATE_ATTRIBUTE_NAME))
+            .sender(mail.getMaybeSender())
+            .recipients(ImmutableList.copyOf(mail.getRecipients()))
+            .content(() -> new ByteArrayInputStream(byteContent.get()))
```
Comment (mbaechler) [inline server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: Don't you think you are implementing MimeMessageInputStreamSource in another way ?

Comment (chibenwa) [reply server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: If MimeMessageInputStreamSource is not part of the craziness ,  then yes let's reuse it.

Comment (mbaechler) [reply server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailAdapter.java]: Well, it's not well implemented but at least having a single bad implementation is better than having several a-bit-less-bad impls IMO, it makes the future changes easier.

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java (hunk 18 above)
Comment (chibenwa) [reply server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java]: Not from an iterator, very sadly...

Maybe we would gain having a `Iterators.toFlux`  method?

File: server/mailrepository/deleted-messages-vault-repository/src/test/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVaultTest.java (hunk 20 above)
Comment (chibenwa) [reply server/mailrepository/deleted-messages-vault-repository/src/test/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVaultTest.java]: Well, we do not have POJO yet to represent this. Introducing one seems like a big refactoring.

In the meantime doing

```
        XMLConfiguration config = new XMLConfiguration();
        config.load(ClassLoaderUtils.getSystemResourceAsSharedStream("mailrepositorystore.xml"));
```

Does the trick...

**Comment by chibenwa** [comment]
https://github.com/linagora/james-project/pull/2202 handles the MailRepository storage part of this PR

File: server/mailrepository/deleted-messages-vault-repository/src/test/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVaultTest.java (hunk 20 above)
Comment (chibenwa) [reply server/mailrepository/deleted-messages-vault-repository/src/test/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVaultTest.java]: https://github.com/linagora/james-project/pull/2204 contributes THE POJO \o/

File: server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java (hunk 18 above)
Comment (mbaechler) [reply server/mailrepository/deleted-messages-vault-repository/src/main/java/org/apache/james/vault/MailRepositoryBasedDeletedMessageVault.java]: ok

**Comment by chibenwa** [comment]
All sub pull requests are merged
